File: gluecodium/src/test/resources/smoke/constants/output/dart/lib/src/smoke/ConstantsInterface.dart
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import 'package:meta/meta.dart';
import 'package:library/src/_library_init.dart' as __lib;
final _smoke_ConstantsInterface_copy_handle = __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Pointer<Void>),
    Pointer<Void> Function(Pointer<Void>)
  >('smoke_ConstantsInterface_copy_handle');
final _smoke_ConstantsInterface_release_handle = __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('smoke_ConstantsInterface_release_handle');
class ConstantsInterface {
  final Pointer<Void> _handle;
  ConstantsInterface._(this._handle);
  void release() => _smoke_ConstantsInterface_release_handle(_handle);
  static final bool boolConstant = true;
  static final int intConstant = -11;
  static final int uintConstant = 4294967295;
  static final double floatConstant = 2.71;
  static final double doubleConstant = -3.14;
  static final String stringConstant = "Foo bar";
  static final ConstantsInterface_StateEnum enumConstant = ConstantsInterface_StateEnum.on;
}
Pointer<Void> smoke_ConstantsInterface_toFfi(ConstantsInterface value) =>
  _smoke_ConstantsInterface_copy_handle(value._handle);
ConstantsInterface smoke_ConstantsInterface_fromFfi(Pointer<Void> handle) =>
  ConstantsInterface._(_smoke_ConstantsInterface_copy_handle(handle));
void smoke_ConstantsInterface_releaseFfiHandle(Pointer<Void> handle) =>
  _smoke_ConstantsInterface_release_handle(handle);
Pointer<Void> smoke_ConstantsInterface_toFfi_nullable(ConstantsInterface value) =>
  value != null ? smoke_ConstantsInterface_toFfi(value) : Pointer<Void>.fromAddress(0);
ConstantsInterface smoke_ConstantsInterface_fromFfi_nullable(Pointer<Void> handle) =>
  handle.address != 0 ? smoke_ConstantsInterface_fromFfi(handle) : null;
void smoke_ConstantsInterface_releaseFfiHandle_nullable(Pointer<Void> handle) =>
  _smoke_ConstantsInterface_release_handle(handle);
enum ConstantsInterface_StateEnum {
    off,
    on
}
// ConstantsInterface_StateEnum "private" section, not exported.
int smoke_ConstantsInterface_StateEnum_toFfi(ConstantsInterface_StateEnum value) {
  switch (value) {
  case ConstantsInterface_StateEnum.off:
    return 0;
  break;
  case ConstantsInterface_StateEnum.on:
    return 1;
  break;
  default:
    throw StateError("Invalid enum value $value for ConstantsInterface_StateEnum enum.");
  }
}
ConstantsInterface_StateEnum smoke_ConstantsInterface_StateEnum_fromFfi(int handle) {
  switch (handle) {
  case 0:
    return ConstantsInterface_StateEnum.off;
  break;
  case 1:
    return ConstantsInterface_StateEnum.on;
  break;
  default:
    throw StateError("Invalid numeric value $handle for ConstantsInterface_StateEnum enum.");
  }
}
void smoke_ConstantsInterface_StateEnum_releaseFfiHandle(int handle) {}
final _smoke_ConstantsInterface_StateEnum_create_handle_nullable = __lib.nativeLibrary.lookupFunction<
    Pointer<Void> Function(Uint32),
    Pointer<Void> Function(int)
  >('smoke_ConstantsInterface_StateEnum_create_handle_nullable');
final _smoke_ConstantsInterface_StateEnum_release_handle_nullable = __lib.nativeLibrary.lookupFunction<
    Void Function(Pointer<Void>),
    void Function(Pointer<Void>)
  >('smoke_ConstantsInterface_StateEnum_release_handle_nullable');
final _smoke_ConstantsInterface_StateEnum_get_value_nullable = __lib.nativeLibrary.lookupFunction<
    Uint32 Function(Pointer<Void>),
    int Function(Pointer<Void>)
  >('smoke_ConstantsInterface_StateEnum_get_value_nullable');
Pointer<Void> smoke_ConstantsInterface_StateEnum_toFfi_nullable(ConstantsInterface_StateEnum value) {
  if (value == null) return Pointer<Void>.fromAddress(0);
  final _handle = smoke_ConstantsInterface_StateEnum_toFfi(value);
  final result = _smoke_ConstantsInterface_StateEnum_create_handle_nullable(_handle);
  smoke_ConstantsInterface_StateEnum_releaseFfiHandle(_handle);
  return result;
}
ConstantsInterface_StateEnum smoke_ConstantsInterface_StateEnum_fromFfi_nullable(Pointer<Void> handle) {
  if (handle.address == 0) return null;
  final _handle = _smoke_ConstantsInterface_StateEnum_get_value_nullable(handle);
  final result = smoke_ConstantsInterface_StateEnum_fromFfi(_handle);
  smoke_ConstantsInterface_StateEnum_releaseFfiHandle(_handle);
  return result;
}
void smoke_ConstantsInterface_StateEnum_releaseFfiHandle_nullable(Pointer<Void> handle) =>
  _smoke_ConstantsInterface_StateEnum_release_handle_nullable(handle);
// End of ConstantsInterface_StateEnum "private" section.

File: gluecodium/src/test/resources/smoke/constants/output/swift/smoke/ConstantsInterface.swift
//
//
import Foundation
public class ConstantsInterface {
    public static let boolConstant: Bool = true
    public static let intConstant: Int32 = -11
    public static let uintConstant: UInt32 = 4294967295
    public static let floatConstant: Float = 2.71
    public static let doubleConstant: Double = -3.14
    public static let stringConstant: String = "Foo bar"
    public static let enumConstant: ConstantsInterface.StateEnum = ConstantsInterface.StateEnum.on
    let c_instance : _baseRef
    init(cConstantsInterface: _baseRef) {
        guard cConstantsInterface != 0 else {
            fatalError("Nullptr value is not supported for initializers")
        }
        c_instance = cConstantsInterface
    }
    deinit {
        smoke_ConstantsInterface_remove_swift_object_from_wrapper_cache(c_instance)
        smoke_ConstantsInterface_release_handle(c_instance)
    }
    public enum StateEnum : UInt32, CaseIterable, Codable {
        case off
        case on
    }
}
internal func getRef(_ ref: ConstantsInterface?, owning: Bool = true) -> RefHolder {
    guard let c_handle = ref?.c_instance else {
        return RefHolder(0)
    }
    let handle_copy = smoke_ConstantsInterface_copy_handle(c_handle)
    return owning
        ? RefHolder(ref: handle_copy, release: smoke_ConstantsInterface_release_handle)
        : RefHolder(handle_copy)
}
extension ConstantsInterface: NativeBase {
    var c_handle: _baseRef { return c_instance }
}
extension ConstantsInterface: Hashable {
    /// :nodoc:
    public static func == (lhs: ConstantsInterface, rhs: ConstantsInterface) -> Bool {
        return lhs.c_handle == rhs.c_handle
    }
    /// :nodoc:
    public func hash(into hasher: inout Hasher) {
        hasher.combine(c_handle)
    }
}
internal func ConstantsInterface_copyFromCType(_ handle: _baseRef) -> ConstantsInterface {
    if let swift_pointer = smoke_ConstantsInterface_get_swift_object_from_wrapper_cache(handle),
        let re_constructed = Unmanaged<AnyObject>.fromOpaque(swift_pointer).takeUnretainedValue() as? ConstantsInterface {
        return re_constructed
    }
    let result = ConstantsInterface(cConstantsInterface: smoke_ConstantsInterface_copy_handle(handle))
    smoke_ConstantsInterface_cache_swift_object_wrapper(handle, Unmanaged<AnyObject>.passUnretained(result).toOpaque())
    return result
}
internal func ConstantsInterface_moveFromCType(_ handle: _baseRef) -> ConstantsInterface {
    if let swift_pointer = smoke_ConstantsInterface_get_swift_object_from_wrapper_cache(handle),
        let re_constructed = Unmanaged<AnyObject>.fromOpaque(swift_pointer).takeUnretainedValue() as? ConstantsInterface {
        smoke_ConstantsInterface_release_handle(handle)
        return re_constructed
    }
    let result = ConstantsInterface(cConstantsInterface: handle)
    smoke_ConstantsInterface_cache_swift_object_wrapper(handle, Unmanaged<AnyObject>.passUnretained(result).toOpaque())
    return result
}
internal func ConstantsInterface_copyFromCType(_ handle: _baseRef) -> ConstantsInterface? {
    guard handle != 0 else {
        return nil
    }
    return ConstantsInterface_moveFromCType(handle) as ConstantsInterface
}
internal func ConstantsInterface_moveFromCType(_ handle: _baseRef) -> ConstantsInterface? {
    guard handle != 0 else {
        return nil
    }
    return ConstantsInterface_moveFromCType(handle) as ConstantsInterface
}
internal func copyToCType(_ swiftClass: ConstantsInterface) -> RefHolder {
    return getRef(swiftClass, owning: false)
}
internal func moveToCType(_ swiftClass: ConstantsInterface) -> RefHolder {
    return getRef(swiftClass, owning: true)
}
internal func copyToCType(_ swiftClass: ConstantsInterface?) -> RefHolder {
    return getRef(swiftClass, owning: false)
}
internal func moveToCType(_ swiftClass: ConstantsInterface?) -> RefHolder {
    return getRef(swiftClass, owning: true)
}
internal func copyToCType(_ swiftEnum: ConstantsInterface.StateEnum) -> PrimitiveHolder<UInt32> {
    return PrimitiveHolder(swiftEnum.rawValue)
}
internal func moveToCType(_ swiftEnum: ConstantsInterface.StateEnum) -> PrimitiveHolder<UInt32> {
    return copyToCType(swiftEnum)
}
internal func copyToCType(_ swiftEnum: ConstantsInterface.StateEnum?) -> RefHolder {
    return copyToCType(swiftEnum?.rawValue)
}
internal func moveToCType(_ swiftEnum: ConstantsInterface.StateEnum?) -> RefHolder {
    return moveToCType(swiftEnum?.rawValue)
}
internal func copyFromCType(_ cValue: UInt32) -> ConstantsInterface.StateEnum {
    return ConstantsInterface.StateEnum(rawValue: cValue)!
}
internal func moveFromCType(_ cValue: UInt32) -> ConstantsInterface.StateEnum {
    return copyFromCType(cValue)
}
internal func copyFromCType(_ handle: _baseRef) -> ConstantsInterface.StateEnum? {
    guard handle != 0 else {
        return nil
    }
    return ConstantsInterface.StateEnum(rawValue: uint32_t_value_get(handle))!
}
internal func moveFromCType(_ handle: _baseRef) -> ConstantsInterface.StateEnum? {
    defer {
        uint32_t_release_handle(handle)
    }
    return copyFromCType(handle)
}
